DetailPanel expand / collapse animation?

Bringing up a [previously requested feature](https://github.com/mbrn/material-table/issues/1566) which never got a resolution:

Would it be possible to have expand/collapse animation when the detail panel is toggled?
Currently is simply appears when toggled on and disappears when toggled off.... doesn't look great.

I have tried using:
`import Collapse from '@material-ui/core/Collapse';`
and wrapping it around the detailpanel output. But it doesn't affect the behavior.

Would this be possible? Any tips?

Hi @oze4 ... I am actually keen to give it a try.
That said, I have to say that even though I am quite well versed with JS/ReactJS, I am not very experienced woking on npm packages. I wanted to try to make some changes to the fork and test things out. I tried changing `package.json` to:
>"@material-table/core": "appukuttan-shailesh/core#branch_2_3_39_mod",
corresponding to: https://github.com/appukuttan-shailesh/core/tree/branch_2_3_39_mod

But when I delete my existing 'node_modules' and try to install all packages (with above updated package.json), I get the error:
> Module not found: Can't resolve '@material-table/core' in '/home/shailesh/gits/projname/subdir/src'

I am obviously doing something wrong. Any tips? Are there any guidelines anywhere that I could follow to get over the npm packaging bits. Thanks.

Thanks @oze4 for the pointers. I am actually fine on the GitHub and ReactJS front. What I very new to is npm packaging and installations.

The part I am stuck with is:
> Make the changes you want to test

Instead of using the demo app (included within material-table-core), I was trying to install material-table-core from my GitHub fork to my current working app.... and I am failing in getting this to work.

I believe I should make changes to the source file (probably 'material-table-core/src/components/m-table-body-row.js' in this case.... still figuring it out). So I [forked](https://github.com/appukuttan-shailesh/core) the GitHub repo. For starters, I should be able to run my existing app by installing 'material-table-core' from my forked repo.

For example, I have a a completely working app that makes use of material-table-core. The package.json is something like this:

```
{
  "name": "project_name",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@material-table/core": "^3.0.16",
    "@material-ui/core": "^4.11.4",
    "@material-ui/icons": "^4.11.2",
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.1.0",
    "@testing-library/user-event": "^12.1.10",
    "axios": "^0.21.1",
    "notistack": "^1.0.8",
    "nunjucks": "^3.2.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-lines-ellipsis": "^0.15.0",
    "react-scripts": "4.0.3",
    "react-slick": "^0.28.1",
    "react-twitter-embed": "^3.0.3",
    "salted-md5": "^4.0.2",
    "showdown": "^1.9.1",
    "showdown-katex": "^0.8.0",
    "slick-carousel": "^1.8.1",
    "web-vitals": "^1.0.1"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```
As you would see, my app uses "3.0.16" and currently works fine.
But now I when I want it to install from my github repo, I change above 'package.json' to 
>"@material-table/core": "github:appukuttan-shailesh/core",

and also delete the existing package-lock.json

But when I do this the app no longer runs on doing `npm start` and I see the error:
> ./src/App.js    
Module not found: Can't resolve '@material-table/core' in 
'/home/shailesh/gits/projname/subdir/src'

If I could somehow learn how to work with local copy of my package or install from my github repo, I would be able to make some attempt on implementing this feature. The JS bits itself do not seem so daunting, but I am totally new to npm packaging bits.

--- a/__tests__/demo/demo-components/index.js
+++ b/__tests__/demo/demo-components/index.js
@@ -317,14 +317,16 @@ export function OneDetailPanel() {
       data={global_data}
       detailPanel={(rowData) => {
         return (
-          <iframe
-            width="100%"
-            height="315"
-            src="https://www.youtube.com/embed/C0DPdy98e4c"
-            frameborder="0"
-            allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
-            allowfullscreen
-          />
+          <div
+            style={{
+              fontSize: 100,
+              textAlign: 'center',
+              color: 'white',
+              backgroundColor: '#43A047',
+            }}
+          >
+            {rowData.name}
+          </div>
         );
       }}
       options={{ showDetailPanelIcon: false }}
@@ -332,6 +334,73 @@ export function OneDetailPanel() {
     />
   );
 }
+
+export function MultipleDetailPanels() {
+    return (
+      <MaterialTable
+        title="Multiple Detail Panels Preview"
+        columns={global_cols}
+        data={global_data}
+        detailPanel={[
+            {
+              tooltip: 'Show Name',
+              render: rowData => {
+                return (
+                  <div
+                    style={{
+                      fontSize: 100,
+                      textAlign: 'center',
+                      color: 'white',
+                      backgroundColor: '#43A047',
+                    }}
+                  >
+                    {rowData.name}
+                  </div>
+                )
+              },
+            },
+            {
+              icon: 'account_circle',
+              tooltip: 'Show Surname',
+              render: rowData => {
+                return (
+                  <div
+                    style={{
+                      fontSize: 101,
+                      textAlign: 'center',
+                      color: 'white',
+                      backgroundColor: '#E53935',
+                    }}
+                  >
+                    {rowData.surname}
+                  </div>
+                )
+              },
+            },
+            {
+              icon: 'favorite_border',
+              openIcon: 'favorite',
+              tooltip: 'Show Both',
+              render: rowData => {
+                return (
+                  <div
+                    style={{
+                      fontSize: 102,
+                      textAlign: 'center',
+                      color: 'white',
+                      backgroundColor: '#FDD835',
+                    }}
+                  >
+                    {rowData.name} {rowData.surname}
+                  </div>
+                )
+              },
+            },
+          ]}
+        onRowClick={(event, rowData, togglePanel) => togglePanel()}
+      />
+    );
+  }
 
 // A little bit of everything
 export function FrankensteinDemo() {

--- a/__tests__/demo/demo.js
+++ b/__tests__/demo/demo.js
@@ -22,6 +22,7 @@ import { render } from 'react-dom';
 import {
   Basic,
   OneDetailPanel,
+  MultipleDetailPanels,
   DefaultOrderIssue,
   TestingNewActionHandlersProp,
   BulkEdit,
@@ -77,6 +78,9 @@ render(
 
     <h1>One Detail Panel</h1>
     <OneDetailPanel />
+
+    <h1>Multiple Detail Panels</h1>
+    <MultipleDetailPanels />
 
     <h1>Editable</h1>
     <EditableCells />

--- a/src/components/m-table-body-row.js
+++ b/src/components/m-table-body-row.js
@@ -481,7 +481,6 @@ export default function MTableBodyRow(props) {
                   )
                 })
               }
-              {/* {props.data.tableData && props.data.tableData.showDetailPanel ? props.data.tableData.showDetailPanel(props.data):null} */}
           </TableCell>
         </TableRow>
       {props.data.tableData.childRows &&

--- a/src/components/m-table-body-row.js
+++ b/src/components/m-table-body-row.js
@@ -5,6 +5,7 @@ import TableRow from '@material-ui/core/TableRow';
 import IconButton from '@material-ui/core/IconButton';
 import Icon from '@material-ui/core/Icon';
 import Tooltip from '@material-ui/core/Tooltip';
+import Collapse from '@material-ui/core/Collapse';
 import PropTypes from 'prop-types';
 import * as CommonValues from '../utils/common-values';
 import { useDoubleClick } from '../utils/hooks/useDoubleClick';
@@ -450,7 +451,6 @@ export default function MTableBodyRow(props) {
       >
         {renderColumns}
       </TableRow>
-      {props.data.tableData && props.data.tableData.showDetailPanel && (
         <TableRow
         // selected={props.index % 2 === 0}
         >
@@ -466,10 +466,24 @@ export default function MTableBodyRow(props) {
             }
             padding="none"
           >
-            {props.data.tableData.showDetailPanel(props.data)}
+              {(typeof props.detailPanel === 'function') &&
+                <Collapse in={Boolean(props.data.tableData && props.data.tableData.showDetailPanel)} timeout="auto" unmountOnExit>
+                  {props.detailPanel(props.data)}
+                </Collapse>
+              }
+              {(typeof props.detailPanel === 'object') &&
+                props.detailPanel.map((panel, index) => {
+                  return (
+                    <Collapse key={index} in={Boolean(props.data.tableData && props.data.tableData.showDetailPanel) && (props.data.tableData.showDetailPanel || '').toString() ===
+                    panel.render.toString()} timeout="auto" unmountOnExit>
+                      {panel.render(props.data)}
+                    </Collapse>
+                  )
+                })
+              }
+              {/* {props.data.tableData && props.data.tableData.showDetailPanel ? props.data.tableData.showDetailPanel(props.data):null} */}
           </TableCell>
         </TableRow>
-      )}
       {props.data.tableData.childRows &&
         props.data.tableData.isTreeExpanded &&
         props.data.tableData.childRows.map((data, index) => {

